worlds/jakanddaxter/Rules.py

```diff
@@ -0,0 +1,222 @@
+import typing
+from BaseClasses import MultiWorld, CollectionState
+from Options import OptionError
+from . import JakAndDaxterWorld
+from .Options import (EnableOrbsanity,
+                      GlobalOrbsanityBundleSize,
+                      PerLevelOrbsanityBundleSize,
+                      FireCanyonCellCount,
+                      MountainPassCellCount,
+                      LavaTubeCellCount,
+                      CitizenOrbTradeAmount,
+                      OracleOrbTradeAmount)
+from .locs import CellLocations as Cells
+from .Locations import location_table
+from .Levels import level_table
+from .regs.RegionBase import JakAndDaxterRegion
+
+
+def set_orb_trade_rule(world: JakAndDaxterWorld):
+    options = world.options
+    player = world.player
+
+    if options.enable_orbsanity == EnableOrbsanity.option_off:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_vanilla(state, player, required_orbs, required_previous_trade))
+    else:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_orbsanity(state, player, required_orbs, required_previous_trade))
+
+
+def recalculate_reachable_orbs(state: CollectionState, player: int) -> None:
+
+    if not state.prog_items[player]["Reachable Orbs Fresh"]:
+
+        # Recalculate every level, every time the cache is stale, because you don't know
+        # when a specific bundle of orbs in one level may unlock access to another.
+        for level in level_table:
+            state.prog_items[player][f"{level} Reachable Orbs".strip()] = (
+                count_reachable_orbs_level(state, player, state.multiworld, level))
+
+        # Also recalculate the global count, still used even when Orbsanity is Off.
+        state.prog_items[player]["Reachable Orbs"] = count_reachable_orbs_global(state, player, state.multiworld)
+        state.prog_items[player]["Reachable Orbs Fresh"] = True
+
+
+def count_reachable_orbs_global(state: CollectionState,
+                                player: int,
+                                multiworld: MultiWorld) -> int:
+
+    accessible_orbs = 0
+    # Cast all regions upfront to access their unique attributes.
+    for region in typing.cast(typing.List[JakAndDaxterRegion], multiworld.get_regions(player)):
+        # Rely on short-circuiting to skip region.can_reach whenever possible.
+        if region.orb_count > 0 and region.can_reach(state):
+            accessible_orbs += region.orb_count
```

Comment:
```suggestion
    for level in level_table:
        accessible_orbs += state.count(f"{level} Reachable Orbs".strip(), player)
```

Reply:
Tested this change with a predefined seed. The suggestion fails Fill with 248 unplaced items. The original code completes Fill successfully.

I think this might be a misunderstanding - `count_reachable_orbs_*` is used by `recalculate_reachable_orbs`, to _update_ the "Reachable Orbs" prog_items. Counting those very prog_items in `count_reachable_orbs_*` makes no sense because its circular reasoning. The point of iterating the list of regions is to provide those updated values based on what you can accomplish in the world - i.e. the regions you can reach with your current move set/power cells/etc.

Reply:
oh i forgor about the region.can_reach when i wrote that